### PR TITLE
HPCC-15618 Systemd implementation

### PR DESCRIPTION
**Pull request by Michael-Gardner**
@xwang2713 Please take a look when you can.

### CODE REVIEW

**Comment by xwang2713** [comment]
Looks good. I tested on Ubuntu 16.04

**Comment by richardkchapman** [comment]
Should probably be squashed down to a single commit.

@mckellyln @jakesmith Can you look too

File: initfiles/bash/etc/systemd/system/thorslaves-launch.sh.in [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+fi
+[[ -z $slaveIps ]] && exit 1
+
+numOfNodes=${#slaveIps[@]}
+
+for ((i=0;i<${#slaveIps[@]};i++)); do
+    for ((c=0;c<${slavespernode};c++)); do
+        __slavePort=$((${THORSLAVEPORT} + (${c} * ${localthorportinc}))) 
+        __slaveNum=$((${i} + 1 + (${c} * ${#slaveIps[@]})))
+        ssh -o LogLevel=QUIET -o StrictHostKeyChecking=no -o BatchMode=yes -i ${SSHidentityfile} ${SSHusername}@${slaveIps[$i]} "/bin/bash -c '@ADMIN_PATH@/thorslaves-exec.sh ${action} thorslave_${instanceName}_${__slaveNum} ${__slavePort} ${__slaveNum} ${THORNAME} ${THORMASTER} ${THORMASTERPORT}'"
```
Comment (jakesmith) [inline initfiles/bash/etc/systemd/system/thorslaves-launch.sh.in]: Serially calling ssh vs previous frunssh is going to be painful (slow).

I don't see "localthor" check here.

File: thorlcr/slave/thslavemain.cpp [hunk 2]
```diff
@@ -296,6 +296,15 @@ void setSlaveAffinity(unsigned processOnNode)
...
 int main( int argc, char *argv[]  )
 {
+    for (unsigned i=0;i<(unsigned)argc;i++) {
+        if (strcmp("--daemon",argv[i])==0 || strcmp("-d",argv[i])==0) {
+            if (daemon(1,0) || write_pidfile(argv[++i])) {
+                perror("Failed to daemonize");
+                return EXIT_FAILURE;
+            }
+            break;
+        }
```
Comment (jakesmith) [inline thorlcr/slave/thslavemain.cpp]: I'm not sure a slave can ever be a daemon (or at least not without other big changes). Slaves are surrogates to thormaster, launched and managed by it.
So I think this code should be remove from thslavemain.cpp

File: dali/sasha/saserver.cpp [hunk 3]
```diff
@@ -257,6 +258,15 @@ static struct CRequestStop : implements IExceptionHandler
 
 int main(int argc, const char* argv[])
 {
+    for (unsigned i=0;i<(unsigned)argc;i++) {
+        if (strcmp("--daemon",argv[i])==0 || strcmp("-d",argv[i])==0) {
+            if (daemon(1,0) || write_pidfile(argv[++i])) {
+                perror("Failed to daemonize");
+                return EXIT_FAILURE;
+            }
+            break;
```
Comment (jakesmith) [inline dali/sasha/saserver.cpp]: This code looks repeated enough to be commoned up into a routine somewhere

**Comment by jakesmith** [comment]
@Michael-Gardner - some initial comments/questions.

I also tested on 16.04.
However, I ran a complete-uninstall.sh , which removed everything, but left all the services running..

@mckellyln - please look also.

**Comment by jakesmith** [comment]
Does this effect the ability to run make install and hpcc-init manually?

**Comment by Michael-Gardner** [comment]
Jake, It won't affect the ability to do a make install and run hpcc-init manually (when it's complete.)  

Todo:
- integration with make install
- compatibility with prerm/postrm scripts
- systemd service script for backupnode
- look at thor slave daemonization

**Comment by mckellyln** [comment]
@Michael-Gardner does this need a rebase ?
I will test.
What cmds are used to start/stop dafilesrv and hpcc-init now ?

**Comment by richardkchapman** [comment]
Seems to have stalled - what are next steps?

**Comment by Michael-Gardner** [comment]
Going to go ahead and take this down until it's ready for review again.  The todo list above is what needs to be worked on.

**Pull request closed without merging**